File: javascript/fibonacci.js
/**
 * Fibonacci Series Implementation in JavaScript
 * Author: Portfolio Project
 * Description: Generates the Fibonacci sequence up to n terms
 */

/**
 * Generate Fibonacci series using iterative approach
 * @param {number} n - Number of terms to generate
 * @returns {number[]} Array containing the Fibonacci sequence
 */
function fibonacciIterative(n) {
    if (n <= 0) return [];
    if (n === 1) return [0];
    
    const sequence = [0, 1];
    for (let i = 2; i < n; i++) {
        sequence.push(sequence[i - 1] + sequence[i - 2]);
    }
    
    return sequence;
}

/**
 * Generate nth Fibonacci number using recursive approach with memoization
 * @param {number} n - The position in Fibonacci sequence (0-indexed)
 * @param {Object} memo - Memoization cache
 * @returns {number} The nth Fibonacci number
 */
function fibonacciRecursive(n, memo = {}) {
    if (n in memo) return memo[n];
    if (n <= 0) return 0;
    if (n === 1) return 1;
    
    memo[n] = fibonacciRecursive(n - 1, memo) + fibonacciRecursive(n - 2, memo);
    return memo[n];
}

/**
 * Generate Fibonacci series using generator function
 * @param {number} n - Number of terms to generate
 * @yields {number} Next Fibonacci number
 */
function* fibonacciGenerator(n) {
    let a = 0, b = 1;
    let count = 0;
    
    while (count < n) {
        yield a;
        [a, b] = [b, a + b];
        count++;
    }
}

/**
 * Generate Fibonacci series using array reduce
 * @param {number} n - Number of terms to generate
 * @returns {number[]} Array containing the Fibonacci sequence
 */
function fibonacciReduce(n) {
    if (n <= 0) return [];
    if (n === 1) return [0];
    
    return Array(n - 2)
        .fill(0)
        .reduce(
            (acc) => {
                acc.push(acc[acc.length - 1] + acc[acc.length - 2]);
                return acc;
            },
            [0, 1]
        );
}

/**
 * Main function to get Fibonacci series
 * @param {number} n - Number of terms
 * @param {string} method - Method to use ('iterative', 'recursive', 'generator', 'reduce')
 * @returns {number[]} Array of Fibonacci numbers
 */
function getFibonacciSeries(n, method = 'iterative') {
    switch (method) {
        case 'iterative':
            return fibonacciIterative(n);
        case 'recursive':
            const memo = {};
            return Array.from({ length: n }, (_, i) => fibonacciRecursive(i, memo));
        case 'generator':
            return [...fibonacciGenerator(n)];
        case 'reduce':
            return fibonacciReduce(n);
        default:
            throw new Error(`Unknown method: ${method}`);
    }
}

// Export for Node.js
if (typeof module !== 'undefined' && module.exports) {
    module.exports = {
        fibonacciIterative,
        fibonacciRecursive,
        fibonacciGenerator,
        fibonacciReduce,
        getFibonacciSeries
    };
}

// Demo (only runs in Node.js)
if (typeof require !== 'undefined' && require.main === module) {
    const n = 10;
    console.log(`Fibonacci Series (first ${n} terms):`);
    console.log(`Iterative: ${getFibonacciSeries(n, 'iterative')}`);
    console.log(`Recursive: ${getFibonacciSeries(n, 'recursive')}`);
    console.log(`Generator: ${getFibonacciSeries(n, 'generator')}`);
    console.log(`Reduce:    ${getFibonacciSeries(n, 'reduce')}`);
}

File: python/fibonacci.py
"""
Fibonacci Series Implementation in Python
Author: Portfolio Project
Description: Generates the Fibonacci sequence up to n terms
"""


def fibonacci_iterative(n):
    """
    Generate Fibonacci series using iterative approach
    
    Args:
        n: Number of terms to generate
        
    Returns:
        List containing the Fibonacci sequence
    """
    if n <= 0:
        return []
    if n == 1:
        return [0]
    
    sequence = [0, 1]
    for i in range(2, n):
        sequence.append(sequence[i-1] + sequence[i-2])
    
    return sequence


def fibonacci_recursive(n, memo=None):
    """
    Generate nth Fibonacci number using recursive approach with memoization
    
    Args:
        n: The position in Fibonacci sequence (0-indexed)
        memo: Dictionary for memoization
        
    Returns:
        The nth Fibonacci number
    """
    if memo is None:
        memo = {}
    
    if n in memo:
        return memo[n]
    
    if n <= 0:
        return 0
    if n == 1:
        return 1
    
    memo[n] = fibonacci_recursive(n-1, memo) + fibonacci_recursive(n-2, memo)
    return memo[n]


def fibonacci_generator(n):
    """
    Generate Fibonacci series using a generator
    
    Args:
        n: Number of terms to generate
        
    Yields:
        Next Fibonacci number
    """
    a, b = 0, 1
    count = 0
    while count < n:
        yield a
        a, b = b, a + b
        count += 1


def get_fibonacci_series(n, method="iterative"):
    """
    Main function to get Fibonacci series
    
    Args:
        n: Number of terms
        method: Method to use ('iterative', 'recursive', 'generator')
        
    Returns:
        List of Fibonacci numbers
    """
    if method == "iterative":
        return fibonacci_iterative(n)
    elif method == "recursive":
        return [fibonacci_recursive(i) for i in range(n)]
    elif method == "generator":
        return list(fibonacci_generator(n))
    else:
        raise ValueError(f"Unknown method: {method}")


if __name__ == "__main__":
    # Demo
    n = 10
    print(f"Fibonacci Series (first {n} terms):")
    print(f"Iterative: {get_fibonacci_series(n, 'iterative')}")
    print(f"Recursive: {get_fibonacci_series(n, 'recursive')}")
    print(f"Generator: {get_fibonacci_series(n, 'generator')}")
